src/hotspot/share/gc/g1/g1YoungGCPostEvacuateTasks.cpp

```diff
@@ -908,24 +828,49 @@ class G1PostEvacuateCollectionSetCleanupTask2::FreeCollectionSetTask : public G1
   }
 };
 
-class G1PostEvacuateCollectionSetCleanupTask2::ResizeTLABsTask : public G1AbstractSubTask {
+class G1PostEvacuateCollectionSetCleanupTask2::ResizeTLABsAndSwapCardTableTask : public G1AbstractSubTask {
   G1JavaThreadsListClaimer _claimer;
+  volatile bool _non_java_threads_claim;
 
   // There is not much work per thread so the number of threads per worker is high.
   static const uint ThreadsPerWorker = 250;
 
 public:
-  ResizeTLABsTask() : G1AbstractSubTask(G1GCPhaseTimes::ResizeThreadLABs), _claimer(ThreadsPerWorker) { }
+  ResizeTLABsAndSwapCardTableTask()
+    : G1AbstractSubTask(G1GCPhaseTimes::ResizeThreadLABs), _claimer(ThreadsPerWorker), _non_java_threads_claim(false)
+  {
+    G1BarrierSet::g1_barrier_set()->swap_global_card_table();
+  }
 
   void do_work(uint worker_id) override {
-    class ResizeClosure : public ThreadClosure {
+    class SwapCardTableClosure : public ThreadClosure {
     public:
+      void do_thread(Thread* thread) {
+        // The global card table references have already been swapped.
+        G1CardTable::CardValue* new_card_table_base = G1CollectedHeap::heap()->card_table_base();
+        G1ThreadLocalData::set_byte_map_base(thread, new_card_table_base);
+      }
+    } swap_cl;
+
+    // We do not expect too many non-Java threads compared to Java threads, so just
+    // let one worker claim that work.
+    if (!_non_java_threads_claim && !Atomic::cmpxchg(&_non_java_threads_claim, false, true, memory_order_relaxed)) {
```

Comment:
Do non-java threads have card-table-base?

Reply:
This code should not be necessary (any more).